docs: added loading pattern (#283)
carbon-design-system/carbon-website: src/pages/patterns/loading-pattern/index.mdx

---
title: Loading
---

<PageDescription>

Loading patterns are used when information takes an extended amount of time to process and appear on screen. Skeleton states and the loading component are two interactions that communicate to users that data is loading and the screen is not frozen.

</PageDescription>

## Skeleton states

> Skeleton states are simplified versions of components used on an initial page load to indicate that the information on the page has not fully loaded yet. They should only appear for one to three seconds, disappearing once the actual content populates the page.

Reviewer: Potential rephrase / missing points:
- Purpose of skeleton state is to improve a webapp's perceived performance by create a first meaningful "paint". This creates an impression of the app loading faster than it actually does
- Skeleton state helps reduce jumping of a page layout as data loads in.

Author: Awesome, added the first point. Thank you!

I don't think the second point is necessary for this section.

> A skeleton state is a stripped-down version of a component, using a `field-01` rectangular block for text. Motion is used in a skeleton state to convey that the page is not stuck loading but rather that data is still being pulled in.

Author: You're right, it's redundant. Adjusted the sentence accordingly!

Never represent toast notifications, overflow menus, dropdown items, modals, and loaders with skeleton states. Elements inside a modal may have a skeleton state, but the modal itself should not.

## Progressive loading

Progressive loading is when a page loads in batches. The simplest view of the page loads first, followed by progressively more detailed batches until the entire viewport has loaded.

The primary batch should show a page's basic structure (the skeleton state versions of the container-based components), data-based text (the skeleton state version of text) and non-data text. Following batches can include images, content outside of the viewport, interactive (action-based) components, and data-based text.

Not all items need a skeleton state and instead can be expressed as negative or white space until they load. For example, a 600 x 600px image can be shown as a 600 x 600px area of white space until the full image loads.

<Row>
<Column colLg={8}>

> ![Example of a dashboard using skeleton states to demonstrate progressive loading.](/images/prog-loading-1.png)

Reviewer 2: Should this set of images be larger since they are full screens? Do we have rules around when images are 12 col vs 8 col?

Author: We don't have strict rules. Typically header/lead images will be 12 col, but outside of that, it's undefined. I can play around with the image sizes here and see how it looks.

Reviewer 2: Seems as if its a "full screen" type image then 12 col is better, where something more zoomed-in might make sense at 8?

Author: Agreed. I just added this to the style guide @janchild and I are putting together.

<Caption>
  Example of a dashboard using skeleton states to demonstrate progressive
  loading
</Caption>

</Column>
</Row>

## Loading components

A loading component is a rotating shape indicating the application is processing.
Use cases include:

- When data processing takes more than a few seconds but no longer than one minute.
- When data from a user input is being saved or submitted.

### Loading

The [loading component](/components/loading/usage) should be used when the entire page is processing. This is often applied after data is submitted or saved by the user.

<Row>
<Column colLg={8}>

![Example of a large loader in an application](/images/larger-loader.png)

<Caption>Example of a large loader in an application</Caption>

</Column>
</Row>

### Inline loading

Use the inline loading component when a single component is processing. For example, when a user restarts a stopped application, an inline loading component indicates the system is processing the restart request.

<Row>
<Column colLg={8}>

![Example of an inline loader](/images/inline-loading-animated.gif)

<Caption>Example of an inline loader</Caption>

</Column>
</Row>

> ## **Load more** button

Reviewer 3: ```suggestion
## Load more
```

---
title: Loading
---

<PageDescription>

> Loading is applied when additional information takes an extended amount of time to process and appear on screen. Skeleton states and the loading component are two interactions that communicate to users that data is currently loading and the screen is not frozen.

Reviewer 3: ```suggestion
Loading patterns are used when information takes an extended amount of time to process and appear on screen. Skeleton states and the loading component are two interactions that communicate to users that data is loading and the screen is not frozen.
```

</PageDescription>

## Skeleton states

> Skeleton states are simplified versions of components used on an initial page load to indicate that the information on the page has not fully loaded yet. They should only appear for 1-3 seconds, disappearing once the actual content populates the page.

Reviewer 3: ```suggestion
Skeleton states are simplified versions of components used on an initial page load to indicate that the information on the page has not fully loaded yet. They should only appear for one to three seconds, disappearing once the actual content populates the page.
```

A skeleton state is a stripped-down version of a component, using a `field-01` rectangular block for text. Motion is used in a skeleton state to convey that the page is not stuck loading but rather that data is still being pulled in.

![Example of a text skeleton state in a data table](images/skeleton.png)

<Caption>Example of a text skeleton state in a data table</Caption>

### Usage

> Only use skeleton states on container-based components (tiles, structured list, etc.), data-based components (data tables, cards, etc.), and data-based text. In most cases, action components (button, input fields, checkboxes, toggle, etc.) do not need to have a skeleton state.

Reviewer 3: ```suggestion
Only use skeleton states on container-based components (e.g. tiles and structured lists), data-based components (e.g. data tables and cards), and data-based text. In most cases, action components (e.g. buttons, input fields, checkboxes, toggles) do not need to have a skeleton state.
```

The primary batch should show a page's basic structure (the skeleton state versions of the container-based components), data-based text (the skeleton state version of text) and non-data text. Following batches can include images, content outside of the viewport, interactive (action-based) components, and data-based text.

Not all items need a skeleton state and instead can be expressed as negative or white space until they load. For example, a 600 x 600px image can be shown as a 600 x 600px area of white space until the full image loads.

<Row>
<Column colLg={8}>

![Example of a dashboard using skeleton states to demonstrate progressive loading.](/images/prog-loading-1.png)

</Column>
</Row>

<Row>
<Column colLg={8}>

![Example of a dashboard using skeleton states to demonstrate progressive loading.](/images/prog-loading-2.png)

</Column>
</Row>

<Row>
<Column colLg={8}>

![Example of a dashboard using skeleton states to demonstrate progressive loading.](/images/prog-loading-3.png)

<Caption>
  Example of a dashboard using skeleton states to demonstrate progressive

>   loading.

Reviewer 3: ```suggestion
  loading
```

</Caption>

</Column>
</Row>

## Loading components

A loading component is a rotating shape indicating the application is processing.

> 

Reviewer 3: ```suggestion
```

Use cases include:

- When data processing takes more than a few seconds but no longer than one minute.

> - When data from a user input is being saved or submitted

Reviewer 3: ```suggestion
- When data from a user input is being saved or submitted.
```

### Loading

The [loading component](/components/loading/usage) should be used when the entire page is processing. This is often applied after data is submitted or saved by the user.

<Row>
<Column colLg={8}>

![Example of a large loader in an application](/images/larger-loader.png)

<Caption>Example of a large loader in an application</Caption>

</Column>
</Row>

### Inline loading

Use the inline loading component when a single component is processing. For example, when a user restarts a stopped application, an inline loading component indicates the system is processing the restart request.

<Row>
<Column colLg={8}>

![Example of an inline loader](/images/inline-loading-animated.gif)

<Caption>Example of an inline loader</Caption>

</Column>
</Row>

> ## 'Load more' button

Reviewer 3: ```suggestion
## **Load more** button
```

> A 'Load more' button should be used to extend a list where there only a small fraction of options are displayed. It can also be used in cases where the list of options is populated via a database. Using 'Load more' allows the data to load in progressive batches.

Reviewer 3: ```suggestion
A **Load more** button should be used to extend a list where there only a small fraction of options are displayed. It can also be used in cases where the list of options is populated via a database. Using **Load more** allows the data to load in progressive batches.
```

<Row>
<Column colLg={8}>

> ![Example of a 'Load more' button](/images/load-more.png)

Reviewer 3: ```suggestion
![Example of a **Load more** button](/images/load-more.png)
```
